### PR TITLE
Add support for decimal logical type

### PR DESCRIPTION
**Pull request by opti**
It adds support for the decimal logical type, so the following Avro schema can be processed correctly:
```json
{
  "type": "bytes",
  "logicalType": "decimal",
  "precision": 4,
  "scale": 2
}
```
However, It's only supported by the Avro v1.11.x. If that's ok, I'm open for suggestions how to better restrict that feature to supported avro version.

https://avro.apache.org/docs/1.11.1/specification/#decimal

### CODE REVIEW

File: lib/avromatic/model/types/decimal_type.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        def name
+          'decimal'
+        end
+
+        def coerce(input)
+          case input
+          when ::NilClass, ::BigDecimal
+            input
+          when ::Rational
+            input.to_d(DEFAULT_PRECISION)
```
Comment (jturkel) [inline lib/avromatic/model/types/decimal_type.rb:33]: Should we be truncating `Float`s too?

File: lib/avromatic/model/types/decimal_type.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+require 'bigdecimal'
+require 'bigdecimal/util'
```
Comment (jturkel) [inline lib/avromatic/model/types/decimal_type.rb:4]: Why is this required?

File: lib/avromatic/model/types/decimal_type.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        def value_classes
+          VALUE_CLASSES
+        end
+
+        def input_classes
+          INPUT_CLASSES
+        end
+
+        def name
+          'decimal'
```
Comment (jturkel) [inline lib/avromatic/model/types/decimal_type.rb:25]: Can we include the `precision` and `scale` in the name?

File: lib/avromatic/model/types/type_factory.rb [hunk 4]
```diff
@@ -26,6 +27,7 @@ module TypeFactory
           'date' => Avromatic::Model::Types::DateType.new,
           'timestamp-micros' => Avromatic::Model::Types::TimestampMicrosType.new,
           'timestamp-millis' => Avromatic::Model::Types::TimestampMillisType.new,
+          'decimal' => Avromatic::Model::Types::DecimalType.new,
```
Comment (jturkel) [inline lib/avromatic/model/types/type_factory.rb:30]: Could we conditionally add this based on `Avro::VERSION` so that we only support decimals in Avro 1.11.0+?

**Review by jturkel — CHANGES_REQUESTED** [review]
Apologies for the delay reviewing this @opti! It would be awesome to include decimal support in Avromatic.

File: lib/avromatic/model/types/decimal_type.rb (hunk 2 above)
Comment (opti) [reply lib/avromatic/model/types/decimal_type.rb:4]: This provides `to_d` method for other types like `Float`, `Integer`, `String`, etc.

File: lib/avromatic/model/types/decimal_type.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+require 'bigdecimal'
+require 'bigdecimal/util'
+require 'avromatic/model/types/abstract_type'
+
+module Avromatic
+  module Model
+    module Types
+      class DecimalType < AbstractType
+        VALUE_CLASSES = [::BigDecimal].freeze
```
[the remark this answers is not in the document]
Comment (opti) [reply lib/avromatic/model/types/decimal_type.rb:11]: My impression was that `VALUE_CLASSES` is for the return type and not the input for the Avro#encode. Would appreciate any details here.

File: lib/avromatic/model/types/decimal_type.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        INPUT_CLASSES = [::Numeric, ::String].freeze
+        DEFAULT_PRECISION = 3
```
[the remark this answers is not in the document]
Comment (opti) [reply lib/avromatic/model/types/decimal_type.rb:13]: I wasn't sure if Rational needs to be explicitly supported at all. It's Numeric, but Avro ruby library [doesn't do any specific coercion](https://github.com/apache/avro/blob/4e1fefca493029ace961b7ef8889a3722458565a/lang/ruby/lib/avro/logical_types.rb#L122) for it, even though they coerce value with `to_d`, which requires precision to be provided.

I can rework it to use `precision` from the schema.

File: lib/avromatic/model/types/decimal_type.rb (hunk 1 above)
Comment (opti) [reply lib/avromatic/model/types/decimal_type.rb:33]: Wouldn't it be better to delegate that to Avro? https://github.com/apache/avro/blob/4e1fefca493029ace961b7ef8889a3722458565a/lang/ruby/lib/avro/logical_types.rb#L180-L200

File: lib/avromatic/model/types/decimal_type.rb [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+          'decimal'
+        end
+
+        def coerce(input)
+          case input
+          when ::NilClass, ::BigDecimal
+            input
+          when ::Rational
+            input.to_d(DEFAULT_PRECISION)
+          when ::Numeric, ::String
```
[the remark this answers is not in the document]
Comment (opti) [reply lib/avromatic/model/types/decimal_type.rb:34]: BigDecimal supports [coercion for strings](https://docs.ruby-lang.org/en/3.2/String.html#method-i-to_d) out of the box and it seems safe enough.

In comparison to `FloatType` and `IntegerType`, Avro's schema accepts `bytes` as the main type for the `logicalType: "decimal"`, which means String should be accepted.

**Comment by opti** [comment]
@jturkel Thank you for the review. I've made updates for some of your comments and provided feedback for others.

File: lib/avromatic/model/types/decimal_type.rb (hunk 1 above)
Comment (opti) [reply lib/avromatic/model/types/decimal_type.rb]: After more consideration it seems that supporting coercion for `Rational` doesn't make a lot of sense here, because eventually it stumbles into the rounding issue anyway. And if I understand correctly the Avro ruby library's approach, they avoid rounding and instead delegate it to the user by raising `Rounding necessary` error when not rounded value provided.

The only benefit of having support for `Rational` coercion is to not raise error when `input.to_d` is called without `precision`, which is [required by specs](https://ruby-doc.org/3.2.1/exts/bigdecimal/Rational.html#method-i-to_d)

**Comment by opti** [comment]
@jturkel any chance for another review?

File: lib/avromatic/model/types/decimal_type.rb (hunk 1 above)
Comment (jturkel) [reply lib/avromatic/model/types/decimal_type.rb]: I agree with your assessment. We're safer following the same approach as Avro Ruby and delegating any rounding decisions to the client. Thus no need to support `Rational`.

File: lib/avromatic/model/types/decimal_type.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (opti) [reply lib/avromatic/model/types/decimal_type.rb]: Thanks for the explanation. That makes sense. Updated value classes to `[::Numeric]`.

File: lib/avromatic/model/types/decimal_type.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (opti) [reply lib/avromatic/model/types/decimal_type.rb]: `String` has been removed from supported inputs.

File: lib/avromatic/model/types/decimal_type.rb [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+        def name
+          "decimal(#{precision}, #{scale})"
+        end
+
+        def coerce(input)
+          case input
+          when ::NilClass, ::BigDecimal
+            input
+          when ::Numeric
```
Comment (jturkel) [inline lib/avromatic/model/types/decimal_type.rb:33]: Can we restrict this conditional to just `Float` and `Integer` since we're not supporting `Rational` or `Complex`? The implementation of `coercible?` will also require a similar change.

**Review by jturkel — CHANGES_REQUESTED** [review]
One small thing I noticed during the final review but otherwise looks great!

File: lib/avromatic/model/types/decimal_type.rb (hunk 8 above)
Comment (opti) [reply lib/avromatic/model/types/decimal_type.rb:33]: @jturkel Should the `VALUE_CLASSES` and `INPUT_CLASSES` be limited to `[BigDecimal, Float, Integer]` as well then?

Comment (jturkel) [reply lib/avromatic/model/types/decimal_type.rb]: `DecimalType#input_classes` should return `[BigDecimal, Float, Integer]` because those are the types that we support as input for decimal attributes in constructors and attribute setters. `DecimalType#value_classes` should return `[BigDecimal]` because decimal attribute values are always coerced to a `BigDecimal` and a decimal attribute getter will always return a `BigDecimal`.

Comment (opti) [reply lib/avromatic/model/types/decimal_type.rb]: Make sense. Updated.
